### ISSUE TITLE
getting UndefinedFunctionError

### ISSUE THREAD

**Issue by slashmili** [RobertDober/earmark_parser#157]
I'm trying to publish the latest changes for my package 
https://github.com/slashmili/ex_syslogger

when I run `mix hex.publish`, it fails creating the docs
```
....
==> ex_syslogger
Compiling 2 files (.ex)
Generated ex_syslogger app
Generating docs...
** (UndefinedFunctionError) function :earmark_parser_string_lexer.string/1 is undefined (module :earmark_parser_string_lexer is not available)
    (earmark_parser 1.4.41) :earmark_parser_string_lexer.string(~c"JsonFormatter is formatter that produces a properly JSON object string where the level, message, node, and metadata are JSON object root properties.")
    (earmark_parser 1.4.41) lib/earmark_parser/helpers/leex_helpers.ex:20: EarmarkParser.Helpers.LeexHelpers.tokenize/2
    (earmark_parser 1.4.41) lib/earmark_parser/helpers/lookahead_helpers.ex:16: EarmarkParser.Helpers.LookaheadHelpers.opens_inline_code/1
    (earmark_parser 1.4.41) lib/earmark_parser/parser.ex:668: EarmarkParser.Parser._inline_or_text?/2
    (earmark_parser 1.4.41) lib/earmark_parser/parser.ex:495: EarmarkParser.Parser._consolidate_para/4
    (earmark_parser 1.4.41) lib/earmark_parser/parser.ex:213: EarmarkParser.Parser._parse/4
    (earmark_parser 1.4.41) lib/earmark_parser/parser.ex:66: EarmarkParser.Parser.lines_to_blocks/3
    (earmark_parser 1.4.41) lib/earmark_parser/parser.ex:59: EarmarkParser.Parser.parse_lines/3
```

is there an issue or I'm doing something wrong 🤔 

---
my env
```
Erlang/OTP 27 [erts-15.0] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1]

Interactive Elixir (1.16.1) - press Ctrl+C to exit (type h() ENTER for help)
```

**Comment by RobertDober** [comment]
**TL;DR** Something strange happens with the compilation order in your project

**Details:**

I am still looking, but something very strange happens when compiling your project, when e.g. changing the ex_doc dependency to my local fork I still cannot debug because of this

```
 mix docs                                                                                                                                                                                                                     [17:44:41]
==> earmark_parser
Generated earmark_parser app
==> ex_doc
Compiling 2 files (.ex)
Generated ex_doc app
==> ex_syslogger
Compiling 2 files (.ex)
    warning: single-quoted strings represent charlists. Use ~c"" if you indeed want a charlist or use "" instead
    │
 96 │     '#Ref' ++ rest = :erlang.ref_to_list(ref)
    │     ~
    │
    └─ lib/ex_syslogger/json_formatter.ex:96:5

Generated ex_syslogger app

17:44:45.001 [notice] Application makeup exited: exited in: Makeup.Application.start(:normal, [])
    ** (EXIT) an exception was raised:
        ** (UndefinedFunctionError) function Makeup.Application.start/2 is undefined (module Makeup.Application is not available)
            (makeup 1.1.2) Makeup.Application.start(:normal, [])
            (kernel 10.0) application_master.erl:295: :application_master.start_it_old/4

17:44:45.006 [notice] Application nimble_parsec exited: :stopped

17:44:45.006 [notice] Application earmark_parser exited: :stopped

17:44:45.006 [notice] Application eex exited: :stopped
** (MatchError) no match of right hand side value: {:error, {:makeup, {:bad_return, {{Makeup.Application, :start, [:normal, []]}, {:EXIT, {:undef, [{Makeup.Application, :start, [:normal, []], []}, {:application_master, :start_it_old, 4, [file: ~c"application_master.erl", line: 295]}]}}}}}}
    (ex_doc 0.35.0) lib/mix/tasks/docs.ex:336: Mix.Tasks.Docs.run/3
    (mix 1.17.2) lib/mix/task.ex:495: anonymous fn/3 in Mix.Task.run_task/5
    (mix 1.17.2) lib/mix/cli.ex:96: Mix.CLI.run_task/2
    /Users/robert/.asdf/installs/elixir/1.17.2-otp-27/bin/mix:2: (file)
```

I did not see anything special in your app but I am not a specialist. Maybe you should ask in the forum for some help?

Right now I do not have the resources to catch a bug I cannot debug (I can reproduce it only when forking your repo, **once**) lateron I always get caught by the message above.

**Comment by RobertDober** [comment]
Update before I close this.

It should not happen that `mix docs` does not correctly compile EarmarkParser, but that's what happenes, maybe that can be fixed in `ex_doc` I do not know. 
Here is what happenes if I compile the deps explicitly

- mix deps.get
- mix deps.compile
- mix docs => Error 

```
Generated ex_syslogger app

18:13:11.570 [notice] Application makeup exited: exited in: Makeup.Application.start(:normal, [])
    ** (EXIT) an exception was raised:
...
```

the full operation is documented below, HTH

```
 mix deps.get                                                                                                                                                                                                                 [18:12:47]
Resolving Hex dependencies...
Resolution completed in 0.01s
Unchanged:
  earmark_parser 1.4.41
  ex_doc 0.34.2
  jason 1.4.3
  makeup 1.1.2
  makeup_elixir 0.16.2
  makeup_erlang 1.0.1
  nimble_parsec 1.4.0
  syslog 1.1.0
* Getting syslog (Hex package)
* Getting ex_doc (Hex package)
* Getting jason (Hex package)
* Getting earmark_parser (Hex package)
* Getting makeup_elixir (Hex package)
* Getting makeup_erlang (Hex package)
* Getting makeup (Hex package)
* Getting nimble_parsec (Hex package)

╭─>robert@narbo ~/gh/forx/ex_syslogger                                                                                                                                                       ❰1.17.2-otp-27➤27.0❱ [± main/f5cd0ba ✓]
╰─> λ mix deps.compile                                                                                                                                                                                                             [18:12:54]
==> earmark_parser
Compiling 2 files (.xrl)
Compiling 1 file (.yrl)
Compiling 3 files (.erl)
Compiling 46 files (.ex)
Generated earmark_parser app
==> nimble_parsec
Compiling 4 files (.ex)
Generated nimble_parsec app
===> Fetching pc v1.15.0
===> Analyzing applications...
===> Compiling pc
===> Analyzing applications...
===> Compiling syslog
===> Compiling c_src/syslog_drv.c
===> Linking /Users/robert/gh/forx/ex_syslogger/_build/dev/lib/syslog/priv/syslog_drv.so
==> jason
Compiling 10 files (.ex)
     warning: Decimal.to_string/1 is undefined (module Decimal is not available or is yet to be defined)
     │
 231 │     [?", decimal.to_string(value), ?"]
     │                  ~
     │
     └─ lib/encoder.ex:231:18: Jason.Encoder.Decimal.encode/2

     warning: Decimal.to_string/2 is undefined (module Decimal is not available or is yet to be defined)
     │
 242 │     [?", decimal.to_string(value, :normal), ?"]
     │                  ~
     │
     └─ lib/encode.ex:242:18: Jason.Encode.struct/4

    warning: Decimal.new/1 is undefined (module Decimal is not available or is yet to be defined)
    │
 94 │         decimal.new(string)
    │                 ~
    │
    └─ lib/decoder.ex:94:17: Jason.Decoder.float_decode_function/1

warning: struct Decimal.Error is undefined (module Decimal.Error is not available or is yet to be defined)
└─ lib/decoder.ex: Jason.Decoder.float_decode_function/1

Generated jason app
==> makeup
Compiling 15 files (.ex)
Generated makeup app
==> makeup_elixir
Compiling 6 files (.ex)
Generated makeup_elixir app
==> makeup_erlang
Compiling 4 files (.ex)
Generated makeup_erlang app
==> ex_doc
Compiling 26 files (.ex)
Generated ex_doc app

╭─>robert@narbo ~/gh/forx/ex_syslogger                                                                                                                                                       ❰1.17.2-otp-27➤27.0❱ [± main/f5cd0ba ✓]
╰─> λ mix docs                                                                                                                                                                                                                     [18:13:08]
Compiling 2 files (.ex)
    warning: single-quoted strings represent charlists. Use ~c"" if you indeed want a charlist or use "" instead
    │
 96 │     '#Ref' ++ rest = :erlang.ref_to_list(ref)
    │     ~
    │
    └─ lib/ex_syslogger/json_formatter.ex:96:5

Generated ex_syslogger app

18:13:11.570 [notice] Application makeup exited: exited in: Makeup.Application.start(:normal, [])
    ** (EXIT) an exception was raised:
        ** (UndefinedFunctionError) function Makeup.Application.start/2 is undefined (module Makeup.Application is not available)
            (makeup 1.1.2) Makeup.Application.start(:normal, [])
            (kernel 10.0) application_master.erl:295: :application_master.start_it_old/4

18:13:11.575 [notice] Application nimble_parsec exited: :stopped

18:13:11.575 [notice] Application earmark_parser exited: :stopped

18:13:11.575 [notice] Application eex exited: :stopped
** (MatchError) no match of right hand side value: {:error, {:makeup, {:bad_return, {{Makeup.Application, :start, [:normal, []]}, {:EXIT, {:undef, [{Makeup.Application, :start, [:normal, []], []}, {:application_master, :start_it_old, 4, [file: ~c"application_master.erl", line: 295]}]}}}}}}
    (ex_doc 0.34.2) lib/mix/tasks/docs.ex:336: Mix.Tasks.Docs.run/3
    (mix 1.17.2) lib/mix/task.ex:495: anonymous fn/3 in Mix.Task.run_task/5
    (mix 1.17.2) lib/mix/cli.ex:96: Mix.CLI.run_task/2
    /Users/robert/.asdf/installs/elixir/1.17.2-otp-27/bin/mix:2: (file)
```
